docs/blog/_posts/2016-12-05-implicit-function-types.md

```diff
@@ -0,0 +1,170 @@
+---
+layout: blog
+title: Implicit Function Types
+author: Martin Odersky
+authorImg: /images/martin.jpg
+---
+
+I just made the first pull request to add _implicit function types_ to
+Scala. I am pretty excited about it, because, citing the explanation
+of the pull request "This is the first step to bring comonadic
+abstraction to Scala". That's quite a mouthful, so I better explain what I
+mean by it.
+
+Let me try to explain the words in this sentence from right to left.
+
+*Scala*: I assume everyone who reads this understands that we mean the
+ programming language, not the opera house.
+
+*Abstraction*: The ability to name a concept and use just the name afterwards.
+
+*Comonadic*: In category theory, a _comonad_ is the dual of a
+_monad_. Roughly speaking, a monad is a way to wrap the result (or:
+outputs) of a computation in some other type. For instance
+`Future[T]` means that the result of type `T` will be produced at
+some later time on demand, or `Option[T]` indicates that the result
+might also be undefined.
+
+Dually, a _comonad_ allows to transform or
+enrich or otherwise manipulate the _inputs_ to a computation.
+The inputs are typically what a computation can access in its
+environment. Interesting tasks that are by nature comonadic are
+
+ - passing configuration data to the parts of a system that need them,
+ - managing capabilities for security critical tasks,
+ - wiring components up with dependency injection,
+ - defining the meanings of operations with type classes,
+ - more generally, passing any sort of context to a computation.
```

Comment:
Monads and comonads exist for the purpose of *composition*. The *first* thing you need to have either of them is a unary type constructor `F[_]`. You chose a different type constructor for the monad and comonad. To illustrate this, I will fix `F[_]` to `ImplicitFunction1[I, ?]` (for fixed `I`).
Now, the comonad instance:
```scala
def extract[A](function: ImplicitFunction1[I, A]): A
```
We're already stuck. We don't have an `I`. You get around this by assuming implicitly one exists in scope, so actually your type constructor is more like:
```scala
type FunctionAndParam[I, A] = (implicit I, ImplicitFunction[I, A])
```
Which explains where the `Store` is coming from. So `ImplicitFunction1[I, ?] is not a comonad. Now, let's see if it's a monad:
```scala
def pure[A](a: A): ImplicitFunction[I, A] = implicit I => a
def flatMap[A](fa: ImplicitFunction[I, A])(
               f: A => ImplicitFunction[I, B]): ImplicitFunction1[I, B]
 = (implicit i: I) => f(fa(i)).apply(i)
```
Remove the `implicit` parts, and this is just the Reader monad. Implicits do not really need much from the type system or foundations, because from another angle they are just the Reader monad with better syntax and no bind method provided (you can add your own, as you can see there, and that won't magically destroy performance). Then, this *is* monadic code and there can be no performance difference between Reader and implicits - assuming the same *usage*, that is. If both uses are closure-heavy there's nothing the compiler can do, and monadic code *does* tend to be heavily abstracted out as well just because of the demographics involved.

Essentially, my objection to that bimonad is that it's just a comonad and monad which already exist with "implicit" pasted at the start, so the monad will not actually teach us anything about the underlying dependency calculus that implicit params capture. All it teaches us is that implicit functions with the same param type compose, like ordinary functions.

Coeffects also *include a model of implicit parameters*, as you can see on the website, which *may* be useful as theoretical underpinnings, but I agree that otherwise the concept is very large (and I'm very excited to see capture-free closures or affine types if you're hinting that we're going that direction :D).